apiserver/facades/client/application/deployrepository.go

```diff
@@ -154,72 +152,100 @@ func (api *DeployFromRepositoryAPI) DeployFromRepository(arg params.DeployFromRe
+// PendingResourceUpload is only returned for local resources
+// which will require the client to upload the resource once
+// the DeployFromRepository returns. Errors are not terminal,
+// and will be collected and returned altogether.
+func (v *deployFromRepositoryValidator) resolveResources(
+	curl *charm.URL,
+	origin corecharm.Origin,
+	deployResArg map[string]string,
+	resMeta map[string]resource.Meta,
+) ([]resource.Resource, []*params.PendingResourceUpload, error) {
 	var pendingUploadIDs []*params.PendingResourceUpload
-	var errs []error
-	pendingIDs := make(map[string]string)
+	var resources []resource.Resource
 
 	for name, meta := range resMeta {
 		r := resource.Resource{
-			Meta: meta,
-		}
-		deployValue, ok := deployRes[name]
-		if !ok {
-			// TODO, this is temporary until resolving resources is completed.
-			// irl, this should never fail.
-			continue
+			Meta:     meta,
+			Origin:   resource.OriginStore,
+			Revision: -1,
 		}
-		if rev, err := strconv.Atoi(deployValue); err == nil {
-			r.Revision = rev
-			r.Origin = resource.OriginStore
-		} else {
+		deployValue, ok := deployResArg[name]
+		if ok {
+			// resource flag is used on the cli, either a resource revision, or a filename
+			if providedRev, err := strconv.Atoi(deployValue); err == nil {
+				// a resource revision is provided
+				r.Revision = providedRev
+				resources = append(resources, r)
+				continue
+			}
+			// a file is coming from the client
 			r.Origin = resource.OriginUpload
+
+			// add a PendingResourceUpload for this resource to be uploaded by the client
+			pendingUploadIDs = append(pendingUploadIDs, &params.PendingResourceUpload{
+				Name:     meta.Name,
+				Type:     meta.Type.String(),
+				Filename: deployValue,
+			})
+			resources = append(resources, r)
+			continue
 		}
+		resources = append(resources, r)
+	}
+
+	repo, err := v.getCharmRepository(origin.Source)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	resolvedResources, resolveErr := repo.ResolveResources(resources, corecharm.CharmID{URL: curl, Origin: origin})
+
+	return resolvedResources, pendingUploadIDs, resolveErr
+}
+
+// addPendingResource adds a pending resource doc for all resources to be
+// added when deploying the charm. All resources will be
+// processed. Errors are not terminal. It also returns the name to pendingIDs
+// map that's needed by the AddApplication.
+func (api *DeployFromRepositoryAPI) addPendingResources(appName string, resources []resource.Resource) (map[string]string, []error) {
+	var errs []error
+	pendingIDs := make(map[string]string)
+
+	for _, r := range resources {
 		pID, err := api.state.AddPendingResource(appName, r)
 		if err != nil {
-			logger.Warningf("Unable to add pending resource %v for application %v", name, appName)
+			logger.Warningf("Unable to add pending resource %v for application %v", r.Name, appName)
```

Comment:
Should be a deployRepoLogger.Errorf and add the error message too to answer why.